### PR TITLE
Add bgsubtract0r plugin.

### PR DESCRIPTION
**Pull request by smimram**
Frei0r plugin to perform background subtraction on a still video.

This plugin operates on still videos, i.e. the camera is not moving, and its aim is to remove the background image, keeping only the moving parts. It is thus a way to have a green screen without a green screen. The first image is taken as reference for the background, so the video should start without the moving stuff, who should come in afterwards.

A few settings can be set such as:
- the threshold in order to decide whether a pixel is the same as in the reference image
- whether to remove "noise" (i.e. isolated pixels)
- optional blurring of edges

Some recommendations:
- obviously the background should be of a (really) different color than the moving actor
- light in order to avoid shadows on the background

### CODE REVIEW

File: src/filter/bgsubtract0r/bgsubtract0r.c [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+  inst->threshold = 26;
+  inst->reference = NULL;
+  inst->mask = malloc(width*height);
+  return (f0r_instance_t)inst;
+}
+
+void f0r_destruct(f0r_instance_t instance)
+{
+  bgsubtract0r_instance_t* inst = (bgsubtract0r_instance_t*)instance;
+  if (inst->reference) free(inst->reference);
```
Comment (ddennedy) [inline src/filter/bgsubtract0r/bgsubtract0r.c]: It is not necessary to check a pointer before calling free() on it.

File: src/filter/bgsubtract0r/bgsubtract0r.c [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+#define MAX(x,y) ((x) > (y) ? (x) : (y));
+
+inline static uint8_t dst(uint32_t x, uint32_t y)
+{
+  uint8_t d;
+
+  d = abs((x&0xff)-(y&0xff));
+  d = MAX(d, abs(((x>>8)&0xff)-((y>>8)&0xff)));
+  d = MAX(d, abs(((x>>16)&0xff)-((y>>16)&0xff)));
+
```
Comment (ddennedy) [inline src/filter/bgsubtract0r/bgsubtract0r.c]: These operations are not endian-safe. It is better to pass in a const uint8_t* and use array indexing to access the color components.

File: src/filter/bgsubtract0r/bgsubtract0r.c [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+          if (n<=2) mask[width*j+i] = 0;
+        }
+        else
+        {
+          if (n>=6) mask[width*j+i] = 0xff;
+        }
+      }
+
+  for (i=0; i<len; i++)
+    outframe[i] = inframe[i] & 0xffffff | mask[i] << 24;
```
Comment (ddennedy) [inline src/filter/bgsubtract0r/bgsubtract0r.c]: This also is not endian-safe.

File: src/filter/bgsubtract0r/bgsubtract0r.c [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+          {
+            ii = i+di;
+            jj = j+dj;
+            if (ii < 0 || ii >= width || jj < 0 || jj >= height)
+              a += 0xff;
+            else
+              a += mask[width*jj+ii];
+          }
+        a /= s;
+        outframe[width*j+i] = (outframe[width*j+i] & 0xffffff) | a << 24;
```
Comment (ddennedy) [inline src/filter/bgsubtract0r/bgsubtract0r.c]: not endian-safe

File: src/filter/bgsubtract0r/bgsubtract0r.c (hunk 3 above)
Comment (ddennedy) [inline src/filter/bgsubtract0r/bgsubtract0r.c]: Consider something like?
uint8_t* p;
...for...
p = &outframe[i];
p[3] = mask[i];

**Comment by smimram** [comment]
Thanks for your comments, I'll integrate those!

**Comment by smimram** [comment]
Done.

**Pull request merged**